Improve self hosting documentation (#1508)
encoredev/encore: docs/selfhosting/configure-infra.md

---
title: Configure Infrastructure
seotitle: Configure Infrastructure
seodesc: Learn how to configure infrastructure resources for your Encore app.
---

If you are using infrastructure resources, such as SQL databases, Pub/Sub, or metrics, you will need to configure your Docker image with the necessary configuration.
The `build` command lets you provide this by specifying a path to a config file using the `--config` flag.

```bash
encore build docker --config path/to/infra-config.json MY-IMAGE:TAG
```

The configuration file should be a JSON file using the [Encore Infra Config](https://encore.dev/schemas/infra.schema.json) schema.

> This includes configuring things like:

Reviewer: Maybe "supports" is better to indicate not everything is required?

```suggestion
This supports configuring things like:
```

- How to access infrastructure resources (what provider to use, what credentials to use, etc.)
- How to call other services over the network ("service discovery"),
  most notably their base URLs.
- Observability configuration (where to export metrics, etc.)
- Metadata about the environment the application is running in, to power Encore's [metadata APIs](/docs/develop/metadata).
- The values for any application-defined secrets.

This configuration is necessary for the application to behave correctly.

## Example

Here's an example configuration file you can use.

```json
{
  "$schema": "https://encore.dev/schemas/infra.schema.json",
  "metadata": {
    "app_id": "my-app",
    "env_name": "my-env",
    "env_type": "production",
    "cloud": "gcp",
    "base_url": "https://my-app.com"
  },
  "sql_servers": [
    {
      "host": "my-db-host:5432",
      "databases": {
        "my-db": {
          "username": "my-db-owner",
          "password": {"$env": "DB_PASSWORD"}
        }
      }
    }
  ],
  "service_discovery": {
    "myservice": {

>       "base_url": "https://my-service:8044"

Reviewer: Probably better to use the same hostname as the service name, since that's the most common setup?

```suggestion
      "base_url": "https://myservice:8044"
```

See below for examples of each type of infrastructure resource.

### 1. Basic Environment Metadata Configuration

```json
{
  "metadata": {
    "app_id": "my-encore-app",
    "env_name": "production",
    "env_type": "production",
    "cloud": "aws",
    "base_url": "https://api.myencoreapp.com"
  }
}
```

- `app_id`: The ID of your Encore application.
- `env_name`: The environment name, such as `production`, `staging`, or `development`.
- `env_type`: Specifies the type of environment (`production`, `test`, `development`, or `ephemeral`).
- `cloud`: The cloud provider hosting the infrastructure (e.g., `aws`, `gcp`, or `azure`).
- `base_url`: The base URL for services in the environment.

### 2. Graceful Shutdown Configuration

```json
{
  "graceful_shutdown": {
    "total": 30,
    "shutdown_hooks": 10,
    "handlers": 20
  }
}
```

- `total`: The total time allowed for the shutdown process in seconds.
- `shutdown_hooks`: The time allowed for executing shutdown hooks.
- `handlers`: The time allocated for processing request handlers during the shutdown.

### 3. Authentication Methods Configuration

> Private endpoints will not require authentication if no authentication methods are specified. To secure private endpoints, you can configure authentication methods.

Reviewer: ```suggestion
Private endpoints will not require authentication if no authentication methods are specified. This is typically fine when services are deployed on a private network such as a VPC. But sometimes you might need to connect to other services over the public internet, in which case you'll want to ensure private endpoints are only accessible to other backend services. To do that you can configure authentication methods.
```
